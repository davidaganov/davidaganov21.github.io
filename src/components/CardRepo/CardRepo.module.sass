.card
  display: flex
  flex-direction: column
  border: .2rem solid var(--color-accent)
  padding: 2.5rem 2rem
  transition: all .2s
  &:hover
    transform: translateY(-.15rem)
    box-shadow: .35rem .35rem 0 .05rem var(--color-accent)
  @media (min-width: 769px)
    height: 23rem

.title
  text-transform: capitalize
  font: bold 2rem var(--main-font)
  color: var(--color-accent)

.description
  margin: 1rem 0 2rem
  color: var(--color-white-700)
  @media (min-width: 769px)
    word-wrap: anywhere
    display: -webkit-box
    -webkit-line-clamp: 2
    -webkit-box-orient: vertical
    overflow: hidden

.bottom
  margin-top: auto

.link
  margin-left: -.2rem
  &--live
    position: relative
    margin-left: 1.8rem
    &::before
      position: absolute
      content: ""
      width: .6rem
      height: .6rem
      left: -1.3rem
      top: .1rem
      bottom: 0
      margin: auto 0
      border-radius: 1rem
      background-color: var(--color-white-700)

.tags
  display: flex
  flex-wrap: wrap
  gap: 1rem
  padding-top: 1.5rem
  li
    padding: .7rem
    text-transform: uppercase
    font: normal .8rem var(--second-font)
    border: .15rem solid var(--color-white-500)
    color: var(--color-white-500)
