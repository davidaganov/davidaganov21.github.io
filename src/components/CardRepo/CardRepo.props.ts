import { DetailedHTMLProps, HTMLAttributes } from "react"
import { ReposProps } from "../../interfaces"

export interface CardRepoProps
  extends DetailedHTMLProps<HTMLAttributes<HTMLLIElement>, HTMLLIElement> {
  card: ReposProps
}
