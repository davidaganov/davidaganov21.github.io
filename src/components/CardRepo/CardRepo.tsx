import { useTranslation } from "react-i18next"
import { CardRepoProps } from "./CardRepo.props"
import styles from "./CardRepo.module.sass"
import cn from "classnames"

export const CardRepo = ({ card, className }: CardRepoProps) => {
  const { name, description, html_url, topics, homepage } = card
  const clearName = name.replace(/[.\-/\\\s]/g, " ")

  const { t } = useTranslation()

  const renderTags = () => {
    const items = topics.map((tag, i) => {
      return (
        <li
          key={i}
          lang="en"
        >
          {tag}
        </li>
      )
    })

    return <ul className={styles.tags}>{items}</ul>
  }

  const renderDemo = () => {
    if (homepage) {
      return (
        <a
          href={homepage}
          className={cn(styles.link, styles["link--live"], "inline-link")}
          target="_blank"
          rel="noreferrer"
          aria-label={`${t("repositories.go_demo")} ${clearName}`}
        >
          {t("repositories.go_demo")}
          <span
            lang="en"
            className="visually-hidden"
          >
            {clearName}
          </span>
        </a>
      )
    }
  }

  const tags = renderTags()
  const demo = renderDemo()

  return (
    <li className={cn(styles.card, className)}>
      <h3
        className={styles.title}
        lang="en"
      >
        {clearName}
      </h3>
      <p
        className={styles.description}
        lang="en"
      >
        {description}
      </p>
      <div className={styles.bottom}>
        <a
          href={html_url}
          className={cn(styles.link, "inline-link")}
          target="_blank"
          rel="noreferrer"
        >
          {t("repositories.go_repo")}
          <span
            lang="en"
            className="visually-hidden"
          >
            {clearName}
          </span>
        </a>
        {demo}
        {tags}
      </div>
    </li>
  )
}
