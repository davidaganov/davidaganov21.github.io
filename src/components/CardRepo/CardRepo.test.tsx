import { render, screen } from "@testing-library/react"
import { CardRepo } from "./CardRepo"

const data = {
  id: "1",
  name: "Test",
  description: "Description project",
  html_url: "repo",
  topics: ["first", "second"],
  homepage: "demo"
}

describe("CardRepo component", () => {
  it("CardRepo renders", () => {
    render(<CardRepo card={data} />)

    expect(screen.getByText(/test/i)).toBeInTheDocument()
  })

  it("There is a link to live demo", () => {
    render(<CardRepo card={data} />)

    expect(screen.getAllByRole("link")).toHaveLength(2)
  })

  it("No link to live demo", () => {
    render(<CardRepo card={{ ...data, homepage: "" }} />)

    expect(screen.getAllByRole("link")).toHaveLength(1)
  })

  it("CardRepo snapshot", () => {
    const card = render(<CardRepo card={data} />)

    expect(card).toMatchSnapshot()
  })
})
