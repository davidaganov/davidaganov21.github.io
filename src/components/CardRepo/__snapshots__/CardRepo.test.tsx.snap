// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardRepo component CardRepo snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="project"
      >
        <h3
          class="title"
        >
          Test
        </h3>
        <p
          class="description"
        >
          Description project
        </p>
        <div
          class="bottom"
        >
          <a
            aria-label="Go to repository Test"
            class="link inline-link"
            href="repo"
            rel="noreferrer"
            target="_blank"
          >
            Go to repository
          </a>
          <a
            aria-label="Live demo Test"
            class="link link--live inline-link"
            href="demo"
            rel="noreferrer"
            target="_blank"
          >
            Live demo
          </a>
          <ul
            class="tags"
          >
            <li>
              first
            </li>
            <li>
              second
            </li>
          </ul>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="project"
    >
      <h3
        class="title"
      >
        Test
      </h3>
      <p
        class="description"
      >
        Description project
      </p>
      <div
        class="bottom"
      >
        <a
          aria-label="Go to repository Test"
          class="link inline-link"
          href="repo"
          rel="noreferrer"
          target="_blank"
        >
          Go to repository
        </a>
        <a
          aria-label="Live demo Test"
          class="link link--live inline-link"
          href="demo"
          rel="noreferrer"
          target="_blank"
        >
          Live demo
        </a>
        <ul
          class="tags"
        >
          <li>
            first
          </li>
          <li>
            second
          </li>
        </ul>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
